### PR TITLE
[dist, parallel] fix: preserve shared gather gradients

### PR DESCRIPTION
**Pull request by 0z5a**
### What does this PR do?

Fix `_Gather.backward` modifying gradients shared with another autograd branch, and address the full-clone memory overhead raised in review.

On NCCL, nonempty real gradients with positive shard sizes use reduce-scatter. Borrowed inputs get a separate output; required owned packing may reuse the local rank's slice where the prior all-reduce result also retained full storage. Other backends and complex/empty cases keep an owned contiguous all-reduce buffer. Scaling stays before summation; the no-sum path scales only the local slice.

This is a bounded shared-primitive fix related to #993, not completion of USP/FSDP validation.

### Checklist Before Starting

- Related: #993 and the [coordination/evidence comment](https://github.com/ByteDance-Seed/VeOmni/issues/993#issuecomment-5561651057).
- Inspected related #1111, #1153, #1158, #961, #1044 and #1087.
- Addresses the [memory review](https://github.com/ByteDance-Seed/VeOmni/pull/1159#discussion_r3947042028) and [test-consolidation review](https://github.com/ByteDance-Seed/VeOmni/pull/1159#discussion_r3947118409).
- Title follows the repository format; no breaking API changes.

#### Latest merge/conflict resolution (2026-09-08)

Current head: `e597ed4815d78d165ee6c141db21e92718c7059b`, merging upstream main `e706d620ab0c872a833e3f8359bb388d2a05e70c` into the existing PR without rewriting history.

- Resolve the single conflict in `.agents/knowledge/constraints.md`: retain upstream's optional MagiAttention-extra installation instructions and the PR's backward-gradient-buffer ownership constraint. No gather implementation or test changes were needed for the resolution.
- `ulysses.py` and `test_all_gather.py` are byte-identical to the previously tested PR head `264590d0d6f3287883cfa02371c82d15c6282c8b`. Other imported source files, including generated models, match upstream main.
- Before and after the merge commit, `make quality` passes with Ruff 0.13.1: lint passes and **596 files** pass format checks. Agent-document path/reference validation and staged/unstaged `git diff --check` pass. Repository-required independent review reports **safe**.
- GitHub confirms the new head is **mergeable with no conflicts**. This does not mean all required CI or approvals have passed. Distributed CPU/GPU/NPU tests were **not rerun for this documentation-only resolution**; the exact-source historical results and hardware limitations below remain attributed to their original commits.

#### Prior CI and HCCL follow-up at 264590d (2026-09-08)

- `3023ab770198a8cc00ffa2bcba07dd3b79121413` fixes the device-API CI failure using the existing production `is_nccl_backend(dist.get_backend(ctx.group))` helper and `dist.Backend.NCCL` constants in the tests. All three existing scanner commands (tasks, veomni, tests) pass; no whitelist or workflow relaxation. The production backend/gradient behavior is preserved.
- `264590d0d6f3287883cfa02371c82d15c6282c8b` adds HCCL only to the shared-gradient regression, guarded by NPU/device/HCCL availability and at least two devices. The worker binds its NPU before process-group initialization. One two-rank spawn retains all 16 layout/sum/scale combinations and all three exact value/ownership assertions. The complex/empty edge-case matrix remains Gloo/NCCL.
- Exact CI-only source: **36 CPU/Gloo tests passed**, zero failures/errors/skips (20.900 s pytest wall time). Exact final source: **36 passed, 1 skipped**, zero failures/errors (21.149 s); the skipped item is HCCL because the validation container has no NPU. Collection confirms shared-gradient backend IDs `gloo`, `nccl`, `hccl`.
- Both runs verify all 917 source entries and actual imports; Torch 2.11.0+cu130 / Transformers 5.9.0, `uv pip check` passes, actual CUDA device count zero, network disabled, read-only source, bounded 8-CPU/16-GiB containers, normal exit 0 and no OOM. These are CPU results, **not new NCCL or HCCL hardware results**.
- Full `make quality` passes with locked Ruff 0.13.2 (591 files), applicable pinned file hooks pass, and separate repository-required independent reviews are safe. No public API/configuration changes or new dependencies.

Current production SHA256: `bd23e842b22a6a3d9d6c015214e78f2d657ddd49b1ea57d860f8a3f762715b7c`; final test SHA256: `6b351ffcce1a3f61f3e93f051e9e283b89638b5f7e1124cf675e250619e66482`.

On an actual supported two-NPU environment, run `pytest -q tests/parallel/ulysses/test_all_gather.py -k 'test_gather_backward_preserves_shared_gradients and hccl'`. HCCL hardware verification and maintainer acceptance remain outstanding.

#### Historical qualification

Historical production source SHA256 at the following commits: `ce6c2283f591bf1cc023342231cb1182bc5b7d608c46c0675033beeed2395e58`.

Earlier test-only follow-up `c9e68c53e525cce109af996b40dc3beb886bf447` addresses [per-case process-start overhead](https://github.com/ByteDance-Seed/VeOmni/pull/1159#discussion_r3949981646). Test SHA256: `546d7730786e4826695ed28735e1c9a53f33983d723a5c226d3740965b7f76a2`.

- One two-rank spawn per backend, with all 16 layout/sum/scale combinations inside the worker. Gloo+NCCL pytest items for this matrix reduce from 32 to 2; no combinations or ownership assertions are removed. Assertion messages include backend, rank, layout and both flags.
- Exact-source Linux CPU: **36 passed**, zero failures/errors/skips (19.994 s JUnit). The changed Gloo item contains all 16 combinations per rank. All 917 archived source entries were verified; only this test file differs from `650b297`.
- Same-host, same-image Gloo testcase time for this matrix: **103.715 s → 6.580 s**, summed from JUnit cases in one baseline/fixed run. This is a test-harness timing observation, not a training-performance benchmark.
- AST audit confirms identical per-combination computation and all three assertions after excluding the new diagnostic labels. Full `make quality` passes with locked Ruff 0.13.2 (591 files), all applicable pinned file hooks pass, and independent review is safe.
- The new batched NCCL invocation has **not** been rerun: the available shared-host GPUs were taken by unrelated active jobs. No new GPU/HCCL pass or CI success is claimed for this test-only commit.

Production-fix qualification at `650b297` before the test-only batching above (historical test SHA256 `9954b21f3890bfcfb4d063f8c0a65bccd493cd93089159f9e059d9f1c25db958`):

- Linux CPU/Gloo: **51 passed**, zero failures/errors/skips, including 22 parallel-state/CP neighbors.
- Published full-clone baseline against the same 12 local-storage regressions: **12 expected failures**; fixed source passes all 12.
- Final two-L4 NCCL/neighbor suite: **19 passed** (17 NCCL regressions + 2 unchanged two-GPU neighbors), zero failures/errors/skips.
- Full `make quality`: **591 files**; repository-required independent review: **safe**.

Repository reproduction (after installing the repository's pinned environment):

```sh
CUDA_VISIBLE_DEVICES="" pytest -q tests/parallel/ulysses/test_all_gather.py tests/parallel/test_parallel_state_registry.py tests/parallel/context_parallel/test_dsv4_cp_sharding.py tests/parallel/context_parallel/test_dsv4_cp_indices.py -k 'not nccl and not AllToAllCommTest'
pytest -q tests/parallel/ulysses/test_all_gather.py -k 'nccl or test_all_gather_shapes_stay_on_host or test_grad_aligned'
```

The separate real two-L4 allocator experiment uses the exact production source, compared with the published full-clone source: **640 numerical/input-ownership checks per rank** (320 baseline + 320 fixed) and **28 memory comparisons per rank**, each with two warmups and three samples. All 168 peak/retained/output-storage non-regression assertions pass across the tested matrix.

For a 64 MiB logical FP32 gradient, max-of-three additional allocated memory:

| Case | Baseline peak / retained MiB | Fixed peak / retained MiB |
|---|---:|---:|
| Equal contiguous dim-0 SUM, unscaled | 64 / 64 | 32 / 32 |
| Equal contiguous dim-0 SUM, scaled | 128 / 64 | 64 / 64 |
| Uneven contiguous dim-0 SUM, unscaled | 64 / 64 | rank 0: 16 / 16; rank 1: 48 / 48 |
| Transposed/narrowed/expanded SUM, unscaled | 64 / 64 | 64 / 64 |

The earlier unpublished v2 increased noncontiguous peak allocation to 96 MiB; owned-pack reuse in this revision removes that regression. Required full packing can still retain full storage; this is not a claim that every output is compact. Interrupted shared-host experiments and two failed L4 instrumentation startups are retained separately and not counted as passing.

Environment: Python 3.12, Torch 2.11.0+cu130, Transformers 5.9.0, NCCL 2.28.9. CPU qualification uses the previously locked GPU-image environment with GPUs hidden. L4 uses a recorded targeted core/NCCL environment (102 packages), **not** the full GPU-extra/attention-kernel stack. Original correctness qualification on two L20s found two shared-gradient failures on main and none after the fix.

**Not covered:** HCCL/NPU hardware, four-rank tests, attention kernels, model training, FSDP parity, end-to-end USP or latency/throughput. HCCL test coverage is now added above; HCCL/NPU hardware execution remains unverified and the review thread is not manually marked resolved.

### API and Usage Example

No public API change. `sum_grad` and `grad_scale` retain their mathematical meaning and scaling order. For `gather(x) + bias`, gather backward preserves the unrelated bias gradient and the caller's incoming buffer.

### Design & Code Changes

- Use equal-shard tensor reduce-scatter without the list API's internal flatten; use the NCCL uneven-list path for unequal positive sizes.
- Materialize rank-local layouts/view bits without choosing inconsistent collectives across ranks.
- Keep owned contiguous fallback storage and scale before reduction.
- Consolidate the matrix in the existing `test_all_gather.py` CI entry point; batch its combinations per backend and add shape/dtype/view/scaling and local-storage regressions.
- Update the backward-buffer constraints. No new host-cleanup behavior or duplicate CI invocation.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [x] Applied the required `make quality` checks.
- [x] Updated documentation and tests in the existing CI entry point.
- [x] No training scripts moved or renamed.
- [x] Repository-required independent code review completed.

AI-assisted implementation and repository-required agent review; no claim of completed human maintainer review.

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed sequence-parallel gradient aggregation modifying gradients shared across computation branches.
  - Preserved correct behavior for varied tensor layouts, uneven partitions, data types, and edge cases.
  - Maintained accurate gradient summation and scaling, including local-only scaling scenarios.
  - Preserved compact local storage where applicable during gradient processing.

- **Tests**
  - Added distributed regression coverage across supported backends, including NCCL and HCCL/NPU environments.
  - Added checks for shared and sibling gradients, local output storage, varied partition sizes, layouts, data types, and gradient configurations.

### CODE REVIEW

File: veomni/distributed/sequence_parallel/ulysses.py [hunk 1]
```diff
@@ -234,23 +234,11 @@ def backward(ctx: Any, grad_output: Tensor) -> Tuple[None, Tensor, None, None, N
...
-        # writes into autograd's own buffer, which is also why it protects against
-        # scribbling in place on a tensor autograd may still own.
-        # ``_GatherConcatSP.backward`` adds a ``.clone()`` for that; reconciling
-        # the two is out of scope here.
-        grad_output = grad_output.contiguous()
         if ctx.sum_grad:
+            # Own the reduction buffer: contiguous() alone aliases an already
+            # contiguous gradient, which autograd may share with another branch.
+            # The explicit format also handles narrowed and stride-zero views.
+            grad_output = grad_output.clone(memory_format=torch.contiguous_format)
```
Comment (FoolPlayer) [inline veomni/distributed/sequence_parallel/ulysses.py:286]: `.clone` will take extra memory usage. It should be fixed.

✅ Confirmed as addressed by @0z5a

File: tests/parallel/ulysses/test_gather_backward.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+# Copyright 2026 Bytedance Ltd. and/or its affiliates
```
Comment (FoolPlayer) [inline tests/parallel/ulysses/test_gather_backward.py]: Can you conbine this test file into `test_all_gather.py`

Comment (0z5a) [reply tests/parallel/ulysses/test_gather_backward.py]: Done in 9fc51289eaba8a7f57b767841e0a1424a2c31d85: the regression is now in test_all_gather.py, and the standalone file and duplicate CI command are removed. The original 32-case matrix is unchanged. CPU/Gloo collection is retained; the merged source passes 38 selected CPU tests and full make quality. Production code is unchanged in this commit. The separate memory-allocation request remains open while I qualify its replacement.

File: veomni/distributed/sequence_parallel/ulysses.py (hunk 1 above)
Comment (0z5a) [reply veomni/distributed/sequence_parallel/ulysses.py:286]: Addressed in 650b297866c8b0cee61c54b5d119c13ce1d1771f. NCCL now reduce-scatters into this rank's output instead of cloning and all-reducing the full gradient. Required owned packing can reuse the rank-local slice where the old result already retained full storage; borrowed inputs never alias the output. The no-sum path scales only its local slice. Other backends and complex/empty cases retain an owned contiguous fallback, with scaling before summation.

Exact production-source validation: 51 CPU/Gloo tests and 19 two-L4 tests (17 NCCL + 2 existing neighbors) passed with zero failures/errors/skips; the 12 local-storage regressions fail as expected on the published full-clone baseline. Full make quality passes over 591 files, and the repository-required independent review is safe.

A separate actual-source two-L4 experiment passed 640 numerical/ownership checks and 28 allocator comparisons per rank (two warmups, three samples per variant). For 64 MiB FP32 gradients, equal contiguous SUM peak/retained allocation drops from 64/64 to 32/32 MiB; scaled SUM peak drops from 128 to 64 MiB. Transposed/narrowed/expanded SUM stays at 64/64 MiB: the unpublished v2's 96 MiB peak regression is removed. All tested peak/retained/output-storage non-regression assertions pass.

The PR body records the matrix, environment, reproduction commands and boundaries. This is not a latency/throughput or full USP/FSDP/model-training result; HCCL/NPU hardware remains unverified. Both L4s have been released and logs are backed up.

@coderabbitai review

File: tests/parallel/ulysses/test_all_gather.py [hunk 3]
```diff
@@ -118,6 +131,162 @@ def test_grad_aligned(self):
...
+@pytest.mark.parametrize("backend", _GATHER_BACKWARD_BACKENDS)
+@pytest.mark.parametrize("layout", ["contiguous", "transposed", "narrowed", "expanded"])
+@pytest.mark.parametrize("sum_grad", [False, True])
+@pytest.mark.parametrize("scale_grad", [False, True])
+def test_gather_backward_preserves_shared_gradients(tmp_path, backend, layout, sum_grad, scale_grad):
+    mp.spawn(
+        _check_gather_backward,
+        args=((tmp_path / "rendezvous").as_uri(), backend, layout, sum_grad, scale_grad),
+        nprocs=2,
+    )
```
Comment (FoolPlayer) [inline tests/parallel/ulysses/test_all_gather.py]: These four `parametrize` layers expand to 32 test items, and every item calls `mp.spawn(nprocs=2)` with its own `init_process_group` / `destroy_process_group`. That is 64 process starts and 32 rendezvous per CI run for a test body that takes milliseconds.

Measured on two ranks (H20, NCCL 2.28, standalone replica of this worker so the numbers isolate the harness):

| structure | Gloo 16 cases | NCCL 16 cases | total |
|---|---:|---:|---:|
| one spawn per case (this PR) | 62.8 s | 101.8 s | **164.6 s** |
| one spawn per backend, matrix looped inside | 3.7 s | 6.1 s | **9.8 s** |

Fixed overhead per case is ~3.9 s on Gloo and ~6.0 s on NCCL, none of it spent on the logic under test. Treat 164.6 s as a lower bound: the replica does not import `veomni`, whereas real spawned children re-import `veomni` and `transformers` under the spawn start method.

`test_gather_backward_edge_cases` further down already uses the cheap shape - a single `parametrize` on backend, with `product(...)` inside the worker running 304 combinations in 2 spawns. Applying that same shape here (parametrize on backend only, move `layout` / `sum_grad` / `scale_grad` into a loop inside `_check_gather_backward`) brings this test from ~165 s to ~10 s. Passing the combination into the assertions via `msg=` keeps failure localization roughly as good as parametrized ids.

One extra consideration now that this lives in `test_all_gather.py`: `npu_unit_tests.yml` runs this file too, so the Gloo half of the matrix adds its ~63 s to the NPU job as well.

**Review by FoolPlayer — COMMENTED** [review]
One comment on test structure only, inline. The consolidation into `test_all_gather.py` and the CPU-collectible guard both look right; the remaining issue is that the matrix test still pays a process spawn per parametrized case.

File: tests/parallel/ulysses/test_all_gather.py (hunk 3 above)
Comment (0z5a) [reply tests/parallel/ulysses/test_all_gather.py]: Addressed in c9e68c53e525cce109af996b40dc3beb886bf447. `test_gather_backward_preserves_shared_gradients` is now parametrized only by backend; each two-rank worker loops over the full 16-case layout/sum/scale product, with one process-group init/destroy. All three ownership/value assertions retain their original inputs and tolerances, and `msg=` records backend, rank, layout, sum and scale for localization. The edge-case worker and production code are unchanged.

On the same CPU-only container image with real VeOmni/Torch imports, the original 16 Gloo items pass. Summed JUnit time for this matrix drops from **103.715 s to 6.580 s** in one baseline/fixed comparison. The fixed selection, including edge/storage and parallel-state/CP neighbors, is **36 passed, zero failures/errors/skips**. An AST audit independently confirms the complete product and identical per-case computation/assertions apart from labels.

Full `make quality` passes over 591 files with locked Ruff 0.13.2, applicable pinned hooks pass, and repository-required independent review is safe. The PR body separates this result from the earlier production qualification. I have not rerun the batched NCCL item because the shared GPU pair is occupied by unrelated active jobs, and have not claimed new NCCL/HCCL coverage. This change is test-only and does not alter the already-qualified production backward.

File: tests/parallel/ulysses/test_all_gather.py [hunk 4]
```diff
@@ -118,6 +119,70 @@ def test_grad_aligned(self):
...
+    "backend",
+    [
+        pytest.param("gloo", marks=pytest.mark.skipif(not dist.is_gloo_available(), reason="Gloo required")),
+        pytest.param(
+            "nccl",
+            marks=pytest.mark.skipif(
+                not IS_CUDA_AVAILABLE or not dist.is_nccl_available() or get_torch_device().device_count() < 2,
+                reason="Two CUDA devices and NCCL required",
+            ),
+        ),
```
[the remark this answers is not in the document]
Comment (0z5a) [reply tests/parallel/ulysses/test_all_gather.py]: Added in 264590d0d6f3287883cfa02371c82d15c6282c8b. The shared-gradient regression now includes guarded HCCL coverage on at least two available NPUs, selects/binds the NPU before process-group initialization, and retains all 16 layout/sum/scale combinations and three exact value/ownership assertions inside one two-rank spawn. The separate complex/empty edge matrix remains Gloo/NCCL.

Also fixed device-api-check in 3023ab770198a8cc00ffa2bcba07dd3b79121413: production uses the existing backend helper; tests use the native Backend.NCCL constant. All three scanner scopes pass without changing the whitelist/workflow.

Exact-source CPU/Gloo validation: CI-only 36 passed/0 skipped; final source 36 passed plus 1 HCCL skip because this isolated container has no NPU, with zero failures/errors. All 917 source entries/imports were checked; backend IDs gloo/nccl/hccl collect correctly. Locked full make quality over 591 files, pinned scoped hooks and separate independent reviews pass. No new NCCL/HCCL hardware execution or maintainer acceptance is claimed. The PR body records reproduction and the remaining NPU hardware boundary. The existing NPU unit workflow already runs this test file, including the new guarded HCCL item. Current-head [NPU unit run 34186262685](https://github.com/ByteDance-Seed/VeOmni/actions/runs/34186262685) is awaiting maintainer approval, with zero jobs at 04:19 UTC; no HCCL hardware pass is claimed yet.

This follow-up is AI-assisted.

**Review by FoolPlayer — APPROVED** [review]
